src/main/java/Duke.java

```diff
@@ -1,10 +1,585 @@
+        if (parts.length >= 3) {
+            boolean isDone = parts[1].equals("X");
+            String description = parts[2];
+            if (parts[0] == "T") { // todo
+                return new ToDo(description, isDone);
+            } else if (parts[0] == "D") { // deadline
+                String deadline = parts[3];
+                return new Deadline(description, deadline, isDone);
+            } else if (parts[0] == "E") { // event
+                String from = parts[3];
+                String to = parts[4];
+                return new Event(description, from, to, isDone);
+            }
+
+        } else {
+            System.out.println("Invalid line format: " + line);
+            return null;
+        }
+        return null;
+    }
+
+    public static void writeTasksToFile(ArrayList<Task> taskList) {
+        try {
+            BufferedWriter writer = new BufferedWriter(new FileWriter(FILE_PATH));
+
+            for (Task task : taskList) {
+                // Write each task to a line in the file
+                writer.write(task.toString());
+                writer.newLine();
+            }
+
+            writer.close();
+        } catch (IOException e) {
+            System.out.println("Error saving tasks to file: " + e.getMessage());
+        }
+    }
+
+
+    /**
+     * Checks with the user if the input is a typo of a command.
+     *
+     * @param input Input collected from the user.
+     * @param command Command to check.
+     * @return True if input is a typo of the command.
+     */
+    public static boolean checkCommandTypo(String input, String command) {
+        if(!input.equals(command)) {
+            signalSays("Did you mean '"+ command + "'? (y/n)");
+            String isCommandCheck = scanner.nextLine();
+            if(isCommandCheck.equals("y")) {
+                return true;
+            } else {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if the input is a permutation of the original.
+     *
+     * @param input Input collected from the user.
+     * @param original String to compare the input to.
+     * @return True if input is a permutation of original.
+     */
+    public static boolean isPermutationMatch(String input, String original) {
+        // Check if user input is a permutation match
+        char[] userInputArray = input.toCharArray();
+        char[] originalArray = original.toCharArray();
+
+        // Sort the arrays to compare
+        java.util.Arrays.sort(userInputArray);
+        java.util.Arrays.sort(originalArray);
+
+        return java.util.Arrays.equals(userInputArray, originalArray);
+    }
+
+    /**
+     * Adds input to the list.
+     *
+     * @param input Input collected from the user.
+     */
+    public static void taskAdded(String input) {
+        Task t = new Task(input, false);
+        taskList.add(t);
+        index += 1;
+        signalSays("Added: " + input);
+    }
+
```

Comment:
Not sure if you are done with the JavaDocs, but it would be great to include them for all the methods. If not, your other JavaDocs look great to me!